lib/hanami/action/application_action.rb

```diff
@@ -122,8 +122,8 @@ def finish(req, res, halted)
         #
         # @since 2.0.0
         # @api public
-        def render?(res)
-          view && res.body.empty?
+        def render?(res, halted)
+          view && res.body.empty? && halted.nil?
```

Comment:
@andrewcroome What do you think about shifting the order of the conditionals around?

```suggestion
          !halted && view && res.body.empty?
```

If we know the action is already halted, feels like we should short circuit from that point, and not even bother with the rest of the checks.

Reply:
Actually wait. There's another alternative, what do you think about this as a reader?

```ruby
        def finish(req, res, halted)
          res.render(view, **req.params) if !halted && render?(res)
          super
        end

        def render?(res)
          view && res.body.empty?
        end
```

This way we keep the `halted` logic contained to `#finish`, and keep `#render?` focus on the view-rendering conditions only.

Reply:
@timriley I think that second option feels good, yes. I've added a commit which adopts it.

Reply:
For the record, I check in with @jodosha and he's happy for us to merge this code as is, which I'll do shortly 😄